account_invoice_fiscal_position_update/models/account_invoice.py

```diff
@@ -27,13 +27,18 @@ def fiscal_position_change(self):
                     account = (
                         product.property_account_income_id or
                         product.categ_id.property_account_income_categ_id)
-                    taxes = product.taxes_id
+                    # M2M fields don't have an option 'company_dependent=True'
+                    # so we need per-company post-filtering
+                    taxes = product.taxes_id.filtered(
+                        lambda tax: tax.company_id == self.company_id)
                 else:
                     account = (
                         product.property_account_expense_id or
                         product.categ_id.property_account_expense_categ_id)
-                    taxes = product.supplier_taxes_id
-                taxes = taxes or account.tax_ids
+                    taxes = product.supplier_taxes_id.filtered(
```

Comment:
Yes, but you can reduce to 2 filtered instead of 3:
```
if inv_type in ('out_invoice', 'out_refund'):
    account = (
       product.property_account_income_id or
       product.categ_id.property_account_income_categ_id)
   taxes = product.taxes_id
else:
    account = (
       product.property_account_expense_id or
       product.categ_id.property_account_expense_categ_id)
   taxes = product.supplier_taxes_id
taxes = taxes.filtered(lambda tax: tax.company_id == self.company_id)
taxes = taxes or account.tax_ids.filtered(
    lambda tax: tax.company_id == self.company_id)
```